x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownMvExpandPastProject.java

```diff
@@ -0,0 +1,112 @@
+                        if (projections.stream()
+                            .anyMatch(
+                                ne -> ne.semanticEquals(alias.child())
+                                    || ne instanceof Alias as && as.child().semanticEquals(alias.child()) && as != alias
+                            ) == false) {
+                            // The alias's original field is not referenced elsewhere, no need to preserve it,
+                            mvExpand = PushDownUtils.resolveRenamesFromProject(mvExpand, pj);
+                            break;
+                        }
+
+                        // for query like: row a = 2 | eval b = a ｜ keep * | mv_expand b
+                        Alias aliasAlias = new Alias(
+                            alias.source(),
+                            TemporaryNameUtils.temporaryName(alias.child(), alias.toAttribute(), 0),
+                            alias.child()
+                        );
+                        projections.set(i, mvExpand.expanded());
+                        pj = new Project(pj.source(), new Eval(aliasAlias.source(), pj.child(), List.of(aliasAlias)), projections);
+                        mvExpand = new MvExpand(mvExpand.source(), pj, aliasAlias.toAttribute(), mvExpand.expanded());
+                        break;
+                    } else if (alias.child().semanticEquals(mvExpand.target().toAttribute())) {
+                        // for query like: row a = 2 | eval b = a ｜ keep * | mv_expand a
+                        Alias aliasAlias = new Alias(
+                            alias.source(),
+                            TemporaryNameUtils.temporaryName(alias.child(), alias.toAttribute(), 0),
+                            alias.child()
+                        );
+                        projections.set(i, alias.replaceChild(aliasAlias.toAttribute()));
+                        pj = new Project(pj.source(), new Eval(aliasAlias.source(), pj.child(), List.of(aliasAlias)), projections);
+                        mvExpand = mvExpand.replaceChild(pj);
+                        break;
+                    }
+                }
+            }
+
+            // Build a map of aliases from the original projection
+            AttributeMap.Builder<Alias> aliasBuilder = AttributeMap.builder();
+            pj.forEachExpression(Alias.class, alias -> aliasBuilder.put(alias.toAttribute(), alias));
+            AttributeMap<Alias> aliases = aliasBuilder.build();
+
+            // Push down the MvExpand past the Project
+            MvExpand pushedDownMvExpand = mvExpand.replaceChild(pj.child());
+
+            // Create a new projection at the top based on mvExpand.output(), plugging back in the aliases
+            List<NamedExpression> newProjections = new ArrayList<>();
+            for (Attribute outputExpr : mvExpand.output()) {
+                Alias alias = aliases.resolve(outputExpr, null);
```

Comment:
We don't normally use `resolve` for an `AttributeMap<Alias>`. It's really meant for `AttributeMap<Attribute>` because it can deal with multiple indirections. We don't do that here, so we should rather just use `get`.